### PR TITLE
test/integration: introduce robust harness and migrate EPP suite

### PR DESCRIPTION
**Pull request by LukeAVanDrie**
**What type of PR is this?**
/kind test
/kind cleanup

**What this PR does / why we need it**:

This PR completes the migration of the integration test suite to the new isolated harness architecture. Following the pattern established in the Body-Based Routing (BBR) migration (#1959), this PR overhauls the EPP hermetic test suite.

The EPP tests are significantly more complex than BBR as they involve the full Kubernetes Controller Runtime lifecycle (Manager, Informers, Caches) and the Scheduler logic. This refactor introduces strict namespace isolation to prevent resource pollution between test cases and promotes server lifecycle logic to shared utilities. This will also be useful for future Flow Control tests which must modify pod metrics to toggle between saturated and unsaturated pool states.

### Key Improvements

1.  Namespace-Scoped Controller Manager:
    *   Introduces `TestHarness`, which generates a unique Namespace (via UUID) for every test case.
    *   Configures the `controller-runtime` Manager with a namespace-scoped cache. This ensures that the EPP server under test only sees Pods and Pools relevant to the current test case, eliminating "ghost" resources from previous runs affecting scheduling decisions.
    *   *Note:* EPP tests currently run serially (`t.Parallel()` is disabled) because they rely on the singleton global Prometheus registry.

2.  Domain-Specific Language (DSL):
    *   Introduces EPP-specific builders in `epp/util_test.go` (e.g., `ReqSubset`, `P()` for pod states) and assertions (e.g., `ExpectRouteTo`, `ExpectBufferResp`).
    *   Reduces visual noise in table-driven tests, making the scheduling logic (Queue vs. KV Cache vs. LoRA Affinity) immediately obvious.

3.  Deterministic Synchronization:
    *   Adds `WaitForReadyPodsMetric`. This blocks test execution until the internal scheduler state (via Prometheus metrics) matches the Kubernetes API state, preventing race conditions where requests are sent before the scheduler has indexed the backend pods.

4.  Shared Infrastructure:
    *   Promotes server lifecycle logic (startup, strict `127.0.0.1` binding, TCP readiness checks) to the shared `test/integration` package.
    *   Updates the existing BBR harness to use these shared utilities, reducing boilerplate.

### Refactoring Note

`test/integration/util.go` still retains some logic specific to the EPP domain. While these belong conceptually in `test/integration/epp/util_test.go`, I opted to keep them in the shared `util.go` package for this PR to minimize the already large diff size.

### Test Coverage Mapping

| Old Test Case Name | New Test Case Name | Changes / Notes |
| :--- | :--- | :--- |
| **Standard Routing Logic** | | |
| `select lower queue and kv cache, no active lora` | `select lower queue and kv cache` | Direct mapping. |
| `select active lora, low queue` | `select active lora, low queue` | Direct mapping. |
| `select lora despite higher kv cache usage` | `select lora despite higher kv cache (affinity)` | Direct mapping. |
| `don't shed requests by default` | `do not shed requests by default` | Direct mapping. |
| **Error Handling & Protocol** | | |
| `invalid json; return body` | `invalid json body` | Direct mapping. |
| `body sent over multiple requests...` | `split body across chunks` | Renamed for clarity. Tests that EPP correctly reassembles fragmented JSON bodies. |
| `no backend pods are available` | `no backend pods available` | Direct mapping. |
| `request don't contains invalid payload, model not exist` | `request missing model field` | Renamed to fix grammar; asserts 400 when JSON is valid but missing mandatory fields. |
| `simple GET Request` | `protocol: simple GET (header only)` | Renamed to clarify intent. |
| **Subsetting & Metadata** | | |
| `select active lora with subsetting tag, all pods available` | `subsetting: select best from subset` | Renamed for clarity. |
| `select active lora with subsetting tag, some pods match` | `subsetting: partial match` | Renamed for clarity. |
| `select active lora with subsetting tag, no pods available` | `subsetting: no pods match` | Renamed for clarity. |
| **Passthrough & Rewrites** | | |
| `inferenceobjective's modelName is not translated, passthrough` | `passthrough: model not in objectives` | Renamed for clarity. |
| `rewrite request model` | `rewrite request model` | Direct mapping. |
| **Response Processing** | | |
| `responsebody sent over multiple requests, content-type is json, buffer` | `response buffering: multi-chunk JSON` | Renamed for clarity. |
| `Response is invalid json; return body` | `response buffering: invalid JSON` | Renamed for clarity. |
| `responsebody sent over a single request...` (JSON variant) | `response buffering: empty EOS chunk (JSON)` | Renamed. Old name was verbose; tests Envoy behavior of sending empty data frame to signal EOS. |
| `responsebody sent over a single request...` (SSE variant) | `response streaming: SSE token counting` | Renamed. Old name was a duplicate copy-paste error; tests token counting logic on SSE streams. |

**Which issue(s) this PR fixes**:
Refactoring groundwork necessary to support complex state-mutation tests (e.g., Flow Control) which require a strictly isolated environment.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by LukeAVanDrie** [comment]
/assign @ahg-g 

I included a test mapping in the PR description which should hopefully make reviewing `hermetic_test.go` a bit easier. It's probably best to review as a full rewrite. I have ensured existing test cases and assertions strictly match.

**Comment by LukeAVanDrie** [comment]
Fixed the boilerplate issue on last push.

**Comment by LukeAVanDrie** [comment]
resolved merge conflict after rebasing

File: test/integration/bbr/harness_test.go [hunk 1]
```diff
@@ -67,58 +60,24 @@ func NewBBRHarness(t *testing.T, ctx context.Context, streaming bool) *BBRHarnes
...
+		t,
+		serverCtx,
+		runnable,
+		port,
+		logger,
+	)
 
 	h := &BBRHarness{
 		t:        t,
 		ctx:      ctx,
```
Comment (shmuelk) [inline test/integration/bbr/harness_test.go]: I realize that this is code from before this PR, but shouldn't the context here be the one created above with the cancel function (i.e. serverCtx) ?

File: test/integration/epp/harness_test.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+		t,
+		serverCtx,
+		runnable,
+		grpcPort,
+		logger,
+	)
+
+	h := &TestHarness{
+		t:            t,
+		ctx:          ctx,
```
Comment (shmuelk) [inline test/integration/epp/harness_test.go]: Shouldn't the context here be serverCtx?

Comment (LukeAVanDrie) [reply test/integration/epp/harness_test.go]: Yes, good catch! Fixed in the latest commit.

File: test/integration/bbr/harness_test.go (hunk 1 above)
Comment (LukeAVanDrie) [reply test/integration/bbr/harness_test.go]: Yes, however, unlike the EPP harness, the BRR harness doesn't need this value right now, so I just removed it from the harness struct entirely in the latest commit.

File: test/integration/epp/hermetic_test.go [hunk 3]
```diff
@@ -14,1351 +14,449 @@ See the License for the specific language governing permissions and
...
-						Key:      requtil.RequestIdHeaderKey,
-						RawValue: []byte("test-request-id"),
-					},
-				},
-			),
 		},
+
+		// --- Error Handling & Edge Cases ----
 		{
-			name: "body sent over multiple requests, noncritical, but one server has capacity, do not shed",
```
Comment (ahg-g) [inline test/integration/epp/hermetic_test.go:339]: where did those tests go? I can't find it if I search for this text.

**Comment by ahg-g** [comment]
Files that are suffixed with `_test` are conventionally files that carry actual tests, not utility functions. In a followup PR, I recommend removing the `_test` suffix for files like `harness_test.go` and `util_test.go` which seems to include utility functions.

File: test/integration/epp/hermetic_test.go (hunk 3 above)
Comment (LukeAVanDrie) [reply test/integration/epp/hermetic_test.go:339]: I included a test mapping table in the PR description.

`body sent over multiple requests...` &rarr; `split body across chunks`

I renamed this for clarity. It tests that EPP correctly reassembles fragmented JSON bodies. The shedding logic is not really the focus here.

**Comment by LukeAVanDrie** [comment]
> Files that are suffixed with `_test` are conventionally files that carry actual tests, not utility functions. In a followup PR, I recommend removing the `_test` suffix for files like `harness_test.go` and `util_test.go` which seems to include utility functions.

I used the `_test.go` suffix intentionally here to utilize Go's build constraints. Since `TestHarness` and these utilities are only used within this package's test suite and are not intended to be exported to other packages, using `_test.go` ensures they are strictly excluded from the production binary and normal build path.

I've seen this convention used differently in different projects (similarly for `_test`-suffixed package names). Happy to change this to match whatever convention we want to uphold for this repo. I agree though, that this can be done in a separate PR (with just the `git mv` commands).

**Comment by ahg-g** [comment]
> > Files that are suffixed with `_test` are conventionally files that carry actual tests, not utility functions. In a followup PR, I recommend removing the `_test` suffix for files like `harness_test.go` and `util_test.go` which seems to include utility functions.
> 
> I used the `_test.go` suffix intentionally here to utilize Go's build constraints. Since `TestHarness` and these utilities are only used within this package's test suite and are not intended to be exported to other packages, using `_test.go` ensures they are strictly excluded from the production binary and normal build path.
> 
> I've seen this convention used differently in different projects (similarly for `_test`-suffixed package names). Happy to change this to match whatever convention we want to uphold for this repo. I agree though, that this can be done in a separate PR (with just the `git mv` commands).

I haven't really seen this convention, I do think using the _test suffix for files other than ones that run tests is surprising and not common, so I recommend renaming in a followup PR. 

I also feel we have too many util test files scattered throughput the repo, we should be able to group them into proper pkgs to better package those functions (perhaps still under a root util folder).
